File: Elixir/apps/day18/lib/day18.ex
defmodule Day18 do
  @moduledoc """
  Documentation for `Day18`.
  """

  require Logger

  def add(num1, nil), do: num1
  def add(nil, num2), do: num2
  def add(num1, num2), do: {num1, num2}

  def split_rec(num) do
    Logger.debug("Split-L #{inspect(num)}")
    {left, right} = num
    cond do
      is_number(left) and left > 9 ->
        half = div(left, 2)
        {{{half, left - half}, right}, true}
      is_tuple(left) ->
        {left_num, did_split} = split_rec(left)
        if did_split do
          {{left_num, right}, true}
        else
          split_rec_right(num)
        end
      true -> split_rec_right(num)
    end
  end

  def split_rec_right(num) do
    Logger.debug("Split-R #{inspect(num)}")
    {left, right} = num
    cond do
      is_number(right) and right > 9 ->
        half = div(right, 2)
        {{left, {half, right - half}}, true}
      is_tuple(right) ->
        {right_num, did_split} = split_rec(right)
        if did_split do
          {{left, right_num}, true}
        else
          {num, false}
        end
      true -> {num, false}
    end
  end

  def split(num) do
    {num, _did_split} = split_rec(num)
    num
  end

  def add_to_left(num, delta) when is_number(num) do
    num + delta
  end

  def add_to_left({left, right}, delta) do
    {add_to_left(left, delta), right}
  end

  def add_to_right(num, delta) when is_number(num) do
    num + delta
  end

  def add_to_right({left, right}, delta) do
    {left, add_to_right(right, delta)}
  end

  def explode_rec({left, right}, 4) do
    {left, right, 0, true}
  end

  def explode_rec(num, depth) do
    Logger.debug("Explode-L #{depth} #{inspect(num)}")
    {left, right} = num
    if is_tuple(left) do
      {left_delta, right_delta, acc, did_explode} = explode_rec(left, depth + 1)
      if did_explode do
        if right_delta do
          {left_delta, nil, {acc, add_to_left(right, right_delta)}, true}
        else
          {left_delta, right_delta, {acc, right}, true}
        end
      else
        explode_rec_right(num, depth)
      end
    else
      explode_rec_right(num, depth)
    end
  end

  def explode_rec_right(num, depth) do
    Logger.debug("Explode-R #{depth} #{inspect(num)}")
    {left, right} = num
    if is_tuple(right) do
      {left_delta, right_delta, acc, did_explode} = explode_rec(right, depth + 1)
      if did_explode do
        if left_delta do
          {nil, right_delta, {add_to_right(left, left_delta), acc}, true}
        else
          {left_delta, right_delta, {left, acc}, true}
        end
      else
        {nil, nil, num, false}
      end
    else
      {nil, nil, num, false}
    end
  end

  def explode(num) do
    {_, _, num, _} = explode_rec(num, 0)
    num
  end

  def reduce(num) do
    Logger.debug("Reduce #{inspect(num)}")
    {_, _, num, did_explode} = explode_rec(num, 0)
    Logger.debug("Exploded #{did_explode}")
    if did_explode do
      reduce(num)
    else
      {num, did_split} = split_rec(num)
      Logger.debug("Split #{did_split}")
      if did_split do
        reduce(num)
      else
        num
      end
    end
  end

  def add_reduce(num1, num2) do
    reduce(add(num1, num2))
  end

  def magnitude(num) when is_number(num), do: num
  def magnitude({left, right}), do: 3 * magnitude(left) + 2 * magnitude(right)

  def to_tuple(n) when is_number(n), do: n
  def to_tuple(list) when is_list(list) do
    list
    |> Enum.map(&to_tuple/1)
    |> List.to_tuple()
  end
  def to_tuple(line) when is_binary(line) do
    {result, _binding} = Code.eval_string(line)
    to_tuple(result)
  end

  def homework() do
    File.read!("data/day18.txt")
    |> String.split("\n")
    |> Stream.filter(&(String.length(&1) > 0))
    |> Stream.map(&to_tuple/1)
    |> Enum.reduce(
         nil,
         fn num, acc ->
           Logger.debug("Add #{inspect(acc)} and #{inspect(num)}")
           add_reduce(acc, num)
         end
       )
    |> magnitude()
  end

  def largest_sum(num1, numbers) do
    Enum.reduce(
      numbers,
      0,
      fn num2, acc ->
        Logger.debug("#{acc} --- Sum #{inspect(num1)} and #{inspect(num2)}")
        max(
          magnitude(add_reduce(num1, num2)),
          max(
            magnitude(add_reduce(num2, num1)),
            acc
          )
        )
      end
    )
  end

  def largest_sum([]), do: 0
  def largest_sum([head | tail]) do
    max(largest_sum(head, tail), largest_sum(tail))
  end

  def backside() do
    File.read!("data/day18.txt")
    |> String.split("\n")
    |> Stream.filter(&(String.length(&1) > 0))
    |> Enum.map(&to_tuple/1)
    |> largest_sum()
  end
end

File: Python/day18.py
#!/usr/bin/env python3
from util import load


def parse(line):
    depth = -1
    num = []
    for c in line:
        if c == '[':
            depth += 1
        elif c == ']':
            depth -= 1
        elif c != ',':
            num.append((int(c), depth))
    return num


def add(num1, num2):
    return [(n, d + 1) for (n, d) in num1] + [(n, d + 1) for (n, d) in num2]


def explode(num):
    idx = next((i for i, (n, d) in enumerate(num) if d == 4), -1)
    if idx >= 0:
        # left
        n, _ = num[idx]
        if idx > 0:
            nl, dl = num[idx - 1]
            num[idx - 1] = nl + n, dl
        # right
        n, _ = num[idx + 1]
        if idx < len(num) - 2:
            nr, dr = num[idx + 2]
            num[idx + 2] = nr + n, dr
        # self
        num[idx:idx + 2] = [(0, 3)]
    return num


def split(num):
    idx = next((i for i, (n, d) in enumerate(num) if n > 9), -1)
    if idx >= 0:
        n, d = num[idx]
        half = n // 2
        num[idx:idx + 1] = [(half, d + 1), (n - half, d + 1)]
    return num


def add_reduce(num1, num2):
    num = add(num1, num2)
    while True:
        length = len(num)
        num = explode(num)
        if length != len(num):
            continue
        num = split(num)
        if length != len(num):
            continue
        break
    return num


def to_tuples(num):
    while len(num) > 1:
        idx = 0
        while idx + 1 < len(num):
            n1, d1 = num[idx]
            n2, d2 = num[idx + 1]
            if d1 == d2:
                num[idx:idx + 2] = [((n1, n2), d1 - 1)]
                break
            idx += 1
    return num[0][0]


def magnitude_r(left, right):
    if type(left) != int:
        left = magnitude(left)
    if type(right) != int:
        right = magnitude(right)
    return 3 * left + 2 * right


def magnitude(num):
    left, right = num
    return magnitude_r(left, right)


def do_homework(lines):
    lhs = parse(lines[0])
    for line in lines[1:]:
        rhs = parse(line)
        lhs = add_reduce(lhs, rhs)
    return magnitude(to_tuples(lhs))


def magnitude_sum(num1, num2):
    return magnitude(to_tuples(add_reduce(num1, num2)))


def do_backside(lines):
    numbers = [parse(line) for line in lines]
    largest = 0
    for i in range(len(numbers)):
        for j in range(i + 1, len(numbers)):
            largest = max(largest, magnitude_sum(numbers[i], numbers[j]))
            largest = max(largest, magnitude_sum(numbers[j], numbers[i]))
    return largest


if __name__ == "__main__":
    assert parse('[[[[[9,8],1],2],3],4]') == [(9, 4), (8, 4), (1, 3), (2, 2), (3, 1), (4, 0)]
    assert add(parse('[1,2]'), parse('[[3,4],5]')) == [(1, 1), (2, 1), (3, 2), (4, 2), (5, 1)]
    assert to_tuples(explode(parse('[[[[[9,8],1],2],3],4]'))) == ((((0, 9), 2), 3), 4)
    assert to_tuples(explode(parse('[[3,[2,[1,[7,3]]]],[6,[5,[4,[3,2]]]]]'))) == \
           ((3, (2, (8, 0))), (9, (5, (4, (3, 2)))))
    assert to_tuples(split([(15, 0), (13, 0)])) == ((7, 8), 13)
    assert to_tuples(add_reduce(parse('[[[[4,3],4],4],[7,[[8,4],9]]]'), parse('[1,1]'))) == \
           ((((0, 7), 4), ((7, 8), (6, 0))), (8, 1))
    assert magnitude(to_tuples(parse('[[[[6,6],[7,6]],[[7,7],[7,0]]],[[[7,7],[7,7]],[[7,8],[9,9]]]]'))) \
           == 4140

    data = load('day18.txt')

    mag = do_homework(data)
    print('Magnitude: %d' % mag)
    assert mag == 2907

    mag = do_backside(data)
    print('Largest magnitude: %d' % mag)
    assert mag == 4690
